### PR TITLE
Adding the Hessian and Laplacian to the NURBS shapefunctions [laplacian_dev]

### PR DESCRIPTION
**Pull request by IdoAkkerman**
The PR adds seconds derivatives  to the NURBS basis. 
This  includes non-uniform weights  and general non-affine mappings.

| PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#586](https://github.com/mfem/mfem/pull/586) | @IdoAkkerman | @tzanio | @kmittal2 + @psocratis | 01/19/20 | 02/16/20 | 02/19/20 | |

### CODE REVIEW

**Comment by tzanio** [comment]
Lets wait on #513 and #476 before assigning this.

**Comment by IdoAkkerman** [comment]
Something must have changed in the build system. Changing changelog broke it...

**Comment by tzanio** [comment]
One travis job was hanging... I restarted it

File: mesh/nurbs.hpp [hunk 1]
```diff
@@ -61,8 +61,10 @@ class KnotVector
 
    int findKnotSpan(double u) const;
 
-   void CalcShape (Vector &shape, int i, double xi) const;
-   void CalcDShape(Vector &grad,  int i, double xi) const;
+   void CalcShape  (Vector &shape, int i, double xi) const;
+   void CalcDShape (Vector &grad,  int i, double xi) const;
+   void CalcDnShape(Vector &gradn, int n, int i, double xi) const;
+   void CalcD2Shape(Vector &grad2, int i, double xi) const { CalcDnShape(grad2, 2, i, xi);}
```
Comment (kmittal2) [inline mesh/nurbs.hpp]: Can you wrap this line since it longer than 80 characters.

File: mesh/nurbs.cpp [hunk 2]
```diff
@@ -211,6 +239,112 @@ void KnotVector::CalcDShape(Vector &grad, int i, double xi) const
...
+      {
+         ndu[j][r] = right[r+1] + left[j-r];
+         temp = ndu[r][j-1]/ndu[j][r];
+         ndu[r][j] = saved + right[r+1]*temp;
+         saved = left[j-r]*temp;
+      }
+      ndu[j][j] = saved;
+   }
+
+   // for (j = 0; j <= p; j++) { gradn[j] = ndu[j][p]; }
```
Comment (kmittal2) [inline mesh/nurbs.cpp]: Do we need this comment?

File: mesh/nurbs.hpp (hunk 1 above)
Comment (IdoAkkerman) [reply mesh/nurbs.hpp]: Yes done.
I think I did perform a "make style" but apparently it did not catch it.
Would it be possible to add this option? Or is it included and I just did not do the "make style"?

File: mesh/nurbs.cpp (hunk 2 above)
Comment (IdoAkkerman) [reply mesh/nurbs.cpp]: No. So removed it.

**Comment by kmittal2** [comment]
The Travis autotest is failing due to code style. Can you please run make style and update the PR?

**Comment by tzanio** [comment]
@IdoAkkerman, note that we are using [Artistic Style](http://astyle.sourceforge.net/) version 2.05.1. Let us know if you have trouble running it, and we will run it for you.

File: miniapps/nurbs/ex1.cpp [hunk 3]
```diff
@@ -43,6 +43,95 @@
...
+   /** Given a particular Finite Element
+       computes the element stiffness matrix elmat. */
+   virtual void AssembleElementMatrix(const FiniteElement &el,
+                                      ElementTransformation &Trans,
+                                      DenseMatrix &elmat)
+   {
+      int nd = el.GetDof();
+      int dim = el.GetDim();
+      int spaceDim = Trans.GetSpaceDim();
+      bool square = (dim == spaceDim);
```
Comment (psocratis) [inline miniapps/nurbs/ex1.cpp]: This seems that it's not used anywhere

**Comment by IdoAkkerman** [comment]
I am having trouble with the "make style" stuff. I am running version 3.1 (standard on ubuntu 18.04).
V 3.1 is also the one offered for download.  So v2.05.1 seems an unavailable downgrade.

**Comment by IdoAkkerman** [comment]
I am also running into compilation errors, which seem to be due to the merging with master.
The errors are:
```
../../libmfem.a(bilininteg.cpp.o):(.data.rel.ro._ZTVN4mfem18GradientIntegratorE[_ZTVN4mfem18GradientIntegratorE]+0x40): undefined reference to `mfem::GradientIntegrator::AssemblePA(mfem::FiniteElementSpace const&, mfem::FiniteElementSpace const&)'
../../libmfem.a(bilininteg.cpp.o):(.data.rel.ro._ZTVN4mfem18GradientIntegratorE[_ZTVN4mfem18GradientIntegratorE]+0x48): undefined reference to `mfem::GradientIntegrator::AddMultPA(mfem::Vector const&, mfem::Vector&) const'
../../libmfem.a(bilininteg.cpp.o):(.data.rel.ro._ZTVN4mfem18GradientIntegratorE[_ZTVN4mfem18GradientIntegratorE]+0x68): undefined reference to `mfem::GradientIntegrator::AddMultTransposePA(mfem::Vector const&, mfem::Vector&) const'
../../libmfem.a(bilininteg.cpp.o):(.data.rel.ro._ZTVN4mfem26VectorDivergenceIntegratorE[_ZTVN4mfem26VectorDivergenceIntegratorE]+0x40): undefined reference to `mfem::VectorDivergenceIntegrator::AssemblePA(mfem::FiniteElementSpace const&, mfem::FiniteElementSpace const&)'
../../libmfem.a(bilininteg.cpp.o):(.data.rel.ro._ZTVN4mfem26VectorDivergenceIntegratorE[_ZTVN4mfem26VectorDivergenceIntegratorE]+0x48): undefined reference to `mfem::VectorDivergenceIntegrator::AddMultPA(mfem::Vector const&, mfem::Vector&) const'
../../libmfem.a(bilininteg.cpp.o):(.data.rel.ro._ZTVN4mfem26VectorDivergenceIntegratorE[_ZTVN4mfem26VectorDivergenceIntegratorE]+0x68): undefined reference to `mfem::VectorDivergenceIntegrator::AddMultTransposePA(mfem::Vector const&, mfem::Vector&) const'
```

**Comment by IdoAkkerman** [comment]
The Hessian and Laplacian are only implemented for the NURBS shape functions and meshes. inline-quad is a standard mesh.

**Comment by psocratis** [comment]
@IdoAkkerman . Oh yes, my bad. Thanks

File: miniapps/nurbs/ex1.cpp (hunk 3 above)
Comment (IdoAkkerman) [reply miniapps/nurbs/ex1.cpp]: Dim is used in one of the integration order computations, but the other 3 can indeed be deleted. 
I will delete when the style and compile issues are resolved.

**Comment by psocratis** [comment]
Compilation goes through for me after merging with master. I still get errors from sample runs though: 

./ex1 -m ../../data/disc-nurbs.mesh -o 1 -no-ibp
Options used:
   --mesh ../../data/disc-nurbs.mesh
   --per none
   --master ''
   --slave ''
   --order '1'
   --no-ibp
   --no-static-condensation
   --visualization
Number of finite element unknowns: 21256


Assertion failed: (data && i >= 0 && i < height && j >= 0 && j < width) is false:
 --> 
 ... in function: double& mfem::DenseMatrix::operator()(int, int)
 ... in file: ../../linalg/densemat.hpp:769

Aborted (core dumped)

**Comment by IdoAkkerman** [comment]
The compile issue have been fixed by the hotfix

**Comment by IdoAkkerman** [comment]
Was also able to downgrade astyle.

**Comment by IdoAkkerman** [comment]
@psocratis You are using a inappropriate discretization for the problem. With --no-ibp the non-integration by parts version of the Laplace problem is solved. This requires a C_1 discretization.
This should give "exactly" the same answer for a formulation with integration by parts,  up to quadrature error. The sole purpose is a thorough check of the Laplacian .

I added some checks to inform this. I also provided 2 meshes that will work (single patch). Also the order should be at least 2.

**Comment by IdoAkkerman** [comment]
Now it should work.
Example:
./nurbs_ex1  -o 2 -no-ibp -m pipe-nurbs-2d.mesh
Options used:
   --mesh pipe-nurbs-2d.mesh
   --per none
   --master ''
   --slave ''
   --order '2'
   --no-ibp
   --no-static-condensation
   --visualization
Number of finite element unknowns: 16900
Size of linear system: 16900
   Iteration :   0  (B r, r) = 0.000843099
.
.
.   Iteration :  71  (B r, r) = 7.10655e-16
Average reduction factor = 0.822186

**Comment by IdoAkkerman** [comment]
I was able to fix the style "error". Unfortunately, the error is non-prescriptive:

=== BEGIN error log from code-style.err
mesh/nurbs.hpp:   void PrintFunctions(std::ostream &out=std::cout, int samples=11) const;
=== END error log from code-style.err

I had to remove the cout default argument. So the function prototype becomes:
 void PrintFunctions(std::ostream &out, int samples=11) const;

This is no big deal, but Ido not see the issue, could this be an astyle/configuration error?

**Comment by tzanio** [comment]
> I was able to fix the style "error". Unfortunately, the error is non-prescriptive:

We have to improve the reporting, and maybe document the checks in `CONTRIBUTING.md`

**Comment by psocratis** [comment]
Hi @IdoAkkerman , 

The sample run looks good now and the checks make it more clear. 
Thanks, 

Socratis

**Comment by tzanio** [comment]
@kmittal2 and @psocratis, reminder that the reviews for this PR are due in less than a week

**Comment by kmittal2** [comment]
Hi @IdoAkkerman , can we also test this new functionality using a 3D example?

**Comment by IdoAkkerman** [comment]
That should work as well. I will add a 3D example.

**Comment by IdoAkkerman** [comment]
It did not work!! But i was able to fix it.

Thank you for the suggestion.

**Comment by kmittal2** [comment]
Also I noticed that only the serial example has been updated. Can we use this new functionality in parallel? If yes, can you update ex1p? If not, can you please indicate in the changelog that the new functionality has been implemented for the serial version only.

**Comment by IdoAkkerman** [comment]
ex1p and ex11p are 2 parallel examples. I believe they work.

File: miniapps/nurbs/ex1p.cpp [hunk 4]
```diff
@@ -1,24 +1,24 @@
 //                     MFEM Example 1 - Parallel NURBS Version
 //
-// Compile with: make ex1p
```
Comment (kmittal2) [inline miniapps/nurbs/ex1p.cpp]: Not quite sure why you changed this name? this should stay ex1p.

File: fem/gridfunc.cpp [hunk 5]
```diff
@@ -475,6 +475,102 @@ const
    GetValues(i, ir, vals, vdim);
 }
 
+void GridFunction::GetLaplacians(int i, const IntegrationRule &ir, Vector &laps,
+                                 int vdim)
+const
+{
+
```
Comment (psocratis) [inline fem/gridfunc.cpp]: unnecessary empty line here

**Comment by tzanio** [comment]
There is still active discussion so I extended the review deadline by a week.

File: fem/fe.cpp [hunk 6]
```diff
@@ -203,6 +203,170 @@ void FiniteElement::CalcPhysDShape(ElementTransformation &Trans,
    Mult(vshape, Trans.InverseJacobian(), dshape);
 }
 
+void FiniteElement::CalcPhysLaplacian(ElementTransformation &Trans,
+                                      Vector &Laplacian) const
+{
+   MFEM_ASSERT(MapType == VALUE, "");
+
+   // Simpler routine if mapping is affine
+   if (Trans.Hessian().FNorm2() < 1e-10)
```
[the remark this answers is not in the document]
Comment (IdoAkkerman) [reply fem/fe.cpp]: Good point. Increased vlaue to 1e-20.

File: miniapps/nurbs/ex1p.cpp (hunk 4 above)
Comment (IdoAkkerman) [reply miniapps/nurbs/ex1p.cpp]: Cmake requires unique executable names.

File: fem/gridfunc.cpp (hunk 5 above)
Comment (IdoAkkerman) [reply fem/gridfunc.cpp]: Removed white space.

**Comment by IdoAkkerman** [comment]
I fixed the memory leak. Which was a small thing in the NURBS class. Not super related in this PR.
But quite non-intrusive.

**Comment by kmittal2** [comment]
@IdoAkkerman the changes are looking good so far. Although I noticed that you changed the names of the examples from `nurbs_ex1` to `ex1` in https://github.com/mfem/mfem/pull/586/commits/c598010352910ec1146f5c1f016b3847ac757a6b and then back to `nurbs_ex1` in https://github.com/mfem/mfem/pull/586/commits/f9235d610a58382d27b5943baf321e120461b1bf.

**Comment by IdoAkkerman** [comment]
Yes, with ex1 cmake gave an error, ex it clashed with the ex1 in the examples directory.
That is why I changed it back.

**Comment by IdoAkkerman** [comment]
@tzanio Comments given on or before Feb 13 I can reply to on the 14th. After that I will be on holiday for a week.

**Comment by kmittal2** [comment]
> Yes, with ex1 cmake gave an error, ex it clashed with the ex1 in the examples directory.
> That is why I changed it back.

Yes, but why do we need to do this in ex1.cpp, ex1p.cpp, and ex11p.cpp? e.g. https://github.com/mfem/mfem/pull/586/files?file-filters%5B%5D=.cpp#diff-d78ef2fc830434c027e1185d92b454bfR3

**Comment by IdoAkkerman** [comment]
> Yes, but why do we need to do this in ex1.cpp, ex1p.cpp, and ex11p.cpp? e.g. https://github.com/mfem/mfem/pull/586/files?file-filters%5B%5D=.cpp#diff-d78ef2fc830434c027e1185d92b454bfR3<

ex1,ex1p an ex11p are executable names that are all ready used in the examples directory. 
CMAKE apperently does not allow an executable name to be used more then once, zo I had to use something else then  ex1,ex1p an ex11p. That is why I opted for the prefix `nurbs_`.

**Comment by IdoAkkerman** [comment]
In my case the current CMakeLists.txt will provide executable with the name nurbs_ex1, nurbs_ex1p and nurbs_ex11p.

**Comment by IdoAkkerman** [comment]
This is the error I get for commit c598010

> CMake Error at config/cmake/modules/MfemCmakeUtilities.cmake:123 (add_executable):
  add_executable cannot create target "ex1" because another target with the
  same name already exists.  The existing target is an executable created in
  source directory "/home/ido/Data/mfem/mfem_laplacian_dev/examples".  See
  documentation for policy CMP0002 for more details.
Call Stack (most recent call first):
  miniapps/nurbs/CMakeLists.txt:24 (add_mfem_miniapp)

**Comment by kmittal2** [comment]
Yes, I agree about the issue with CMake but it is not specific to this PR. Even in Master, we currently use `nurbs_ex1` for the executable name with CMake, but the sample runs in ex1.cpp says `ex1` and not `nurbs_ex1`.  Let's modify ex1.cpp, ex1p.cpp and ex11p.cpp back to what it used to be (https://github.com/mfem/mfem/pull/586/commits/c598010352910ec1146f5c1f016b3847ac757a6b), and we will address this inconsistency later if needed.

**Comment by IdoAkkerman** [comment]
I don't see a problem? Or understand the suggested fix, for that matter.
What is wrong with the 'nurbs_' prefix?
Perhaps it is appropriate to rename the cpp files and also add the 'nurbs_' prefix.

**Comment by IdoAkkerman** [comment]
Commit c598010  does not work!
Hence the change back to state before c598010 which is done by commit f9235d6.
So in my mind there are 3 options:
1) Keep things as is, with prefix and things are working.
2) Remove prefix, but then it does not work.
3) A halfway house were the prefix is only were needed, this leads to inconsistencies though.

If we agree with the nurbs prefix option 1 then there are no inconsistencies to deal with later!
So why not opt for that option.

**Comment by kmittal2** [comment]
What happens if you go to ex1.cpp and change `nurbs_ex1` to `ex1` in Lines 3-11 of ex1.cpp?

**Comment by IdoAkkerman** [comment]
Given that they are all comment lines that would "work". However, the comments would be wrong.
The resuling executable with the current CMakelists.txt is nurbs_ex1, due to the uniqueness issue,
so the current lines 3-11 are the correct, given the current executable name.

I would prevere having correct  comments. When the cmake uniqueness issue is fixed (if ever?) it would be a small fix to also change the comments in the cpp.
Even if the uniqueness isue would be fixed would it be desirable to drop the nurbs_ prefix?
I personally would not really care.

Perhaps it is considered undesirable that the ex1.cpp leads to an nurbs_ex1 executable. This could also be fixed by renaming ex1.cpp into nurbs_ex1.cpp.

File: miniapps/nurbs/ex1p.cpp [hunk 7]
```diff
@@ -1,25 +1,32 @@
...
+//               mpirun -np 4 nurbs_ex1p -m ../../data/square-disc-surf.mesh
+//               mpirun -np 4 nurbs_ex1p -m ../../data/inline-segment.mesh
+//               mpirun -np 4 nurbs_ex1p -m ../../data/amr-quad.mesh
+//               mpirun -np 4 nurbs_ex1p -m ../../data/amr-hex.mesh
+//               mpirun -np 4 nurbs_ex1p -m ../../data/mobius-strip.mesh
+//               mpirun -np 4 nurbs_ex1p -m ../../data/mobius-strip.mesh -o -1 -sc
+//               mpirun -np 4 nurbs_ex1p -m ../../data/square-disc-nurbs.mesh -o -1
+//               mpirun -np 4 nurbs_ex1p -m ../../data/disc-nurbs.mesh -o -1
+//               mpirun -np 4 nurbs_ex1p -m ../../data/pipe-nurbs.mesh -o -1
+//               mpirun -np 4 nurbs_ex1p -m ../../data/beam-hex-nurbs.mesh -pm 1 -ps 2
```
Comment (kmittal2) [inline miniapps/nurbs/ex1p.cpp]: pm and ps are not valid options for ex1p.

**Comment by IdoAkkerman** [comment]
@psocratis thank you for the clarification.
I will correct the `make` route and make sure everything has the `nurbs_` prefix.

File: miniapps/nurbs/ex1p.cpp (hunk 7 above)
Comment (IdoAkkerman) [reply miniapps/nurbs/ex1p.cpp]: I have removed the non-working example.

**Comment by psocratis** [comment]
Hi @IdoAkkerman , 

This is not what I meant. The example files and the sample runs should be named without the prefix as they were before. This is consistent with all other examples throughout the code and with the make build system everything is fine. The only conflict comes from cmake, and therefore the executable files from cmake have to include the prefix.  

Thanks, 

Socratis

**Comment by IdoAkkerman** [comment]
@psocratis The only other miniapp subdir that currently has an example is `performance`. My suggestion would be to "fix" the naming in this miniapp, and add a prefix consistently. Instead of changing the nurbs miniapps back to a halfway-house construction.

**Comment by psocratis** [comment]
Hi @IdoAkkerman, 

I don't have a problem with this. @tzanio what do you think? 

Socratis

File: fem/gridfunc.hpp [hunk 8]
```diff
@@ -152,6 +152,18 @@ class GridFunction : public Vector
    void GetValues(int i, const IntegrationRule &ir, Vector &vals,
                   DenseMatrix &tr, int vdim = 1) const;
 
+   void GetLaplacians(int i, const IntegrationRule &ir, Vector &laps,
```
Comment (tzanio) [inline fem/gridfunc.hpp]: Please document

File: fem/fe.hpp [hunk 9]
```diff
@@ -422,10 +422,29 @@ class FiniteElement
...
    virtual void GetFaceDofs(int face, int **dofs, int *ndofs) const;
 
-   /** each row of h contains the upper triangular part of the hessian
-       of one shape function; the order in 2D is {u_xx, u_xy, u_yy} */
+   /** @brief Evaluate the Hessians of all shape functions of a scalar finite
+       element in reference space at the given point @a ip. */
+   /** Each row of the result DenseMatrix @a Hessian contains upper triangular
+       part of the Hessian of one shape function.
+       The order in 2D is {u_xx, u_xy, u_yy}
+       The size (#Dof x (#Dim (#Dim-1)/2) of @a Hessian must be set in advance.  */
```
Comment (tzanio) [inline fem/fe.hpp]: Thank you for the nice description

File: fem/fe.cpp [hunk 10]
```diff
@@ -203,6 +203,170 @@ void FiniteElement::CalcPhysDShape(ElementTransformation &Trans,
...
+   {
+      for (int nd = 0; nd < Dof; nd++)
+      {
+         Laplacian[nd] = hess(nd,0);
+      }
+   }
+}
+
+
+// Assume a linear mapping
```
Comment (tzanio) [inline fem/fe.cpp]: Is this restriction listed in the doxygen documentation?

**Comment by tzanio** [comment]
I think this is OK as it is for now. @kmittal2 I merged it for testing, but please let me know if you still have any reservations.

**Comment by kmittal2** [comment]
It looks good to me.

**Comment by tzanio** [comment]
The internal regression testing found the following issue. Can you please address?

```
nurbs_ex1.cpp: In member function ‘virtual void Diffusion2Integrator::AssembleElementMatrix(const mfem::FiniteElement&, mfem::ElementTransformation&, mfem::DenseMatrix&)’:
nurbs_ex1.cpp:62:12: warning: unused variable ‘square’ [-Wunused-variable]
       bool square = (dim == spaceDim);
            ^
nurbs_ex1p.cpp: In member function ‘virtual void Diffusion2Integrator::AssembleElementMatrix(const mfem::FiniteElement&, mfem::ElementTransformation&, mfem::DenseMatrix&)’:
nurbs_ex1p.cpp:76:12: warning: unused variable ‘square’ [-Wunused-variable]
       bool square = (dim == spaceDim);
            ^
```

**Comment by tzanio** [comment]
Note: the name changes in this branch cause failures in the `memcheck`, `baseline` scripts on tux. 

We will resolve these once merged in master.

**Comment by kmittal2** [comment]
> The internal regression testing found the following issue. Can you please address?
> 
> ```
> nurbs_ex1.cpp: In member function ‘virtual void Diffusion2Integrator::AssembleElementMatrix(const mfem::FiniteElement&, mfem::ElementTransformation&, mfem::DenseMatrix&)’:
> nurbs_ex1.cpp:62:12: warning: unused variable ‘square’ [-Wunused-variable]
>        bool square = (dim == spaceDim);
>             ^
> nurbs_ex1p.cpp: In member function ‘virtual void Diffusion2Integrator::AssembleElementMatrix(const mfem::FiniteElement&, mfem::ElementTransformation&, mfem::DenseMatrix&)’:
> nurbs_ex1p.cpp:76:12: warning: unused variable ‘square’ [-Wunused-variable]
>        bool square = (dim == spaceDim);
>             ^
> ```

@IdoAkkerman had mentioned he will be on travel so I fixed this and pushed the changes.

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Pull request merged**